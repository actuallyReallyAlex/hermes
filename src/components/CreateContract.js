import React, { useState } from 'react'
import PropTypes from 'prop-types'
import { Box } from 'grommet'
import { itemList } from '../constants'
import { connect } from 'react-redux'
import { createContract } from '../redux/actions/world'
import { generateContract } from '../util'

/**
 * Inputs for creating a contract.
 */
const CreateContract = ({ handleCreateContract, planets }) => {
  const [itemType, setItemType] = useState(itemList[0])
  const [destination, setDestination] = useState({
    location: planets[0].location,
    name: planets[0].name
  })

  return (
    <Box width="medium">
      <label htmlFor="item-type">Item Type</label>
      <select
        id="item-type"
        onChange={e => setItemType(JSON.parse(e.target.value))}
      >
        {itemList.map(item => (
          <option key={item.name} value={JSON.stringify(item)}>
            {item.name}
          </option>
        ))}
      </select>
      <label htmlFor="item-volume">Item Volume</label>
      <input
        disabled
        id="item-volume"
        type="number"
        value={itemType ? itemType.volume : 0}
      />
      <label htmlFor="item-value">Item Value</label>
      <input
        disabled
        id="item-value"
        type="number"
        value={itemType ? itemType.value : 0}
      />
      <label htmlFor="item-destination">Destination</label>
      <select
        id="item-destination"
        onChange={e => {
          const planet = planets.find(planet => planet.id === e.target.value)
          setDestination({ location: planet.location, name: planet.name })
        }}
      >
        {planets.map(planet => (
          <option key={planet.id} value={planet.id}>
            {planet.name}
          </option>
        ))}
      </select>
      <button
        onClick={() => {
          handleCreateContract(
            generateContract(planets, itemType.name, destination)
          )
        }}
      >
        Create Contract
      </button>
    </Box>
  )
}

CreateContract.propTypes = {
  handleCreateContract: PropTypes.func.isRequired,
  planets: PropTypes.array.isRequired
}

const mapStateToProps = ({ world }) => ({ planets: world.planets })

const mapDispatchToProps = dispatch => ({
  handleCreateContract: contract => dispatch(createContract(contract))
})

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(CreateContract)
