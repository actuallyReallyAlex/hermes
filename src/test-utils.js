import React from 'react'
import configureStore from 'redux-mock-store'
import { Provider } from 'react-redux'
import { defaultState } from './fixtures'
import { render } from '@testing-library/react'

const mockThunks = ({ dispatch, getState }) => next => action => {
  const mockedAction = () => ({ type: 'MOCKED ACTION' })

  let returnValue

  if (typeof action === 'function' || action === undefined) {
    returnValue = mockedAction
  } else {
    returnValue = next(action)
  }

  return returnValue
}

const middlewares = [mockThunks]

export const mockStore = configureStore(middlewares)

export const customRender = options => {
  const { component, state, props } = options

  const Component = component

  return render(
    <Provider store={mockStore(state ? state : defaultState)}>
      <Component {...props} />
    </Provider>
  )
}
