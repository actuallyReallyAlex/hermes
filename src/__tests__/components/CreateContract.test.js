import { customRender } from '../../test-utils'
import CreateContract from '../../components/CreateContract'

describe('<CreateContract />', () => {
  it('Should render the <CreateContract /> component.', () => {
    const container = customRender({ component: CreateContract })
    expect(container.asFragment()).toMatchSnapshot()
  })
})
