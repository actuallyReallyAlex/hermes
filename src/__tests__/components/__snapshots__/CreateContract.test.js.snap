// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateContract /> Should render the <CreateContract /> component. 1`] = `
<DocumentFragment>
  <div
    class="StyledBox-sc-13pk1d4-0 dygfSD"
  >
    <label
      for="item-type"
    >
      Item Type
    </label>
    <select
      id="item-type"
    >
      <option
        value="{\\"description\\":\\"Bullets used for shootin'.\\",\\"name\\":\\"Ammunition\\",\\"value\\":3,\\"volume\\":0.8}"
      >
        Ammunition
      </option>
      <option
        value="{\\"description\\":\\"Still used to fight the common cold!\\",\\"name\\":\\"Antibiotics\\",\\"value\\":2,\\"volume\\":0.1}"
      >
        Antibiotics
      </option>
      <option
        value="{\\"description\\":\\"Melted sand.\\",\\"name\\":\\"Glass\\",\\"value\\":1,\\"volume\\":1}"
      >
        Glass
      </option>
      <option
        value="{\\"description\\":\\"Midas' favorite.\\",\\"name\\":\\"Gold\\",\\"value\\":5,\\"volume\\":0.5}"
      >
        Gold
      </option>
      <option
        value="{\\"description\\":\\"Your garden variety ore. From a rock. A space rock.\\",\\"name\\":\\"Ore\\",\\"value\\":2,\\"volume\\":1}"
      >
        Ore
      </option>
      <option
        value="{\\"description\\":\\"Not the kind in your blood, but just as valuable.\\",\\"name\\":\\"Plasma\\",\\"value\\":3,\\"volume\\":0.4}"
      >
        Plasma
      </option>
      <option
        value="{\\"description\\":\\"Shiny!\\",\\"name\\":\\"Silver\\",\\"value\\":4,\\"volume\\":0.5}"
      >
        Silver
      </option>
      <option
        value="{\\"description\\":\\"It's not dirt! It's soil.\\",\\"name\\":\\"Soil\\",\\"volume\\":2,\\"value\\":0.5}"
      >
        Soil
      </option>
      <option
        value="{\\"description\\":\\"Junk! Space junk, to be exact. Prevalent throughout the galaxy.\\",\\"name\\":\\"Space Junk\\",\\"value\\":1,\\"volume\\":0.1}"
      >
        Space Junk
      </option>
      <option
        value="{\\"description\\":\\"Smoke it or chew it. Farmed on distant planets by astronauts in overalls.\\",\\"name\\":\\"Tobacco\\",\\"value\\":3,\\"volume\\":0.2}"
      >
        Tobacco
      </option>
    </select>
    <label
      for="item-volume"
    >
      Item Volume
    </label>
    <input
      disabled=""
      id="item-volume"
      type="number"
      value="0.8"
    />
    <label
      for="item-value"
    >
      Item Value
    </label>
    <input
      disabled=""
      id="item-value"
      type="number"
      value="3"
    />
    <label
      for="item-destination"
    >
      Destination
    </label>
    <select
      id="item-destination"
    >
      <option
        value="0"
      >
        Test Planet 1
      </option>
      <option
        value="1"
      >
        Test Planet 2
      </option>
      <option
        value="2"
      >
        Test Planet 3
      </option>
    </select>
    <button>
      Create Contract
    </button>
  </div>
</DocumentFragment>
`;
